lib/overcommit/hook/pre_push/r_spec.rb

```diff
@@ -1,14 +1,47 @@
 module Overcommit::Hook::PrePush
   # Runs `rspec` test suite before push
+  # .overcommit.yml configuration:
+  # Set `changed: true` to run rspec only over edited files.
+  # Set `remote`: $remote_name` to specify a remote name.
   #
   # @see http://rspec.info/
   class RSpec < Base
     def run
-      result = execute(command)
+      command = ['rspec']
+      remote = config['remote'] || 'origin'
+      result = if config['changed']
+                 return :pass if testable_files(remote).empty?
+                 execute([*command, *testable_files(remote)])
+               else
+                 execute(command)
+               end
       return :pass if result.success?
 
       output = result.stdout + result.stderr
       [:fail, output]
     end
+
+    def testable_files(remote)
+      return [] if Overcommit::GitRepo.modified_file_in_branch(remote).empty?
+      files = Overcommit::GitRepo.modified_file_in_branch(remote).map do |file_path|
+        if test?(file_path) || ruby_file?(file_path)
+          test?(file_path) ? file_path : spec_file(file_path)
+        end
+      end
+      files.uniq.reject { |file| file.nil? || file == false }
+    end
+
+    def test?(file_path)
+      file_path.start_with?('spec/') && file_path.end_with?('_spec.rb')
+    end
+
+    def ruby_file?(file_path)
+      file_path.end_with? '.rb'
+    end
+
+    def spec_file(file_path)
+      spec_path = "spec/#{file_path.insert(-4, '_spec')}".delete('/app')
```

Comment:
`-4` is a big of a magic number. We should base it off of the length of the extension, e.g. `'.rb'.length`.

This code also appears to assume the developer is using a Rails project (from the presence of the `/app` directory). Furthermore, `delete('/app')` is dangerous because it doesn't just delete the first occurrence, e.g.

```ruby
irb(main):001:0> "/app/hello/app-directory".delete("/app")
=> "hello-directory"
```

I don't know the best way to accomplish this, but it might be worth creating a mechanism of defining a "projection". Perhaps using a regex with capture groups to transform source files into their corresponding test files is an approach worth exploring.

Also, you'll want to run tests when the actual test files change. This change appears to only run corresponding test files for changed source files.